0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+Unlike many other programming languages that have added Unicode support over a major version bump,
+Dafny is unusually well-positioned to catch most regressions statically.
+Given this, it does not seem worth the effort to build any additional migration features to help customers adopt this change.
+
+## Compilation/Runtime
+
+This change also has a reasonably small effect on the runtime behavior of any self-contained Dafny program,
+meaning a program that does not pass data over an interface with external code.
+The only difference is that the `char` type will need to be compiled to a native datatype
+that supports 32 bits of data (or technically 21 according to the actual Unicode scalar value range)
+rather than 16 bits as previous versions did.
+Many target languages will have dedicated types designed to hold any Unicode scalar value,
+such as the `char` primitive type in Rust or the `rune` alias for `int32` in Go.
+
+The main complexity in implementing this change is minimizing the migration pain for
+code bases that use `{:extern}` definitions to pass string values in and out of the Dafny runtime.
+This generally requires converting between the internal representation of Dafny strings
+and the native implementation in the target language,
+e.g. between `Dafny.ISequence<char>` and `System.String` in C#.
+Even there, the various Dafny runtime libraries provide helper methods that can be adjusted 
+to work with the new semantics of Unicode characters.
+In the C# runtime, for example, the `Sequence.FromString(string)` method converts a native string
+to a equivalent `Dafny.ISequence<char>` copy.
+A parallel method named something similar to `Sequence.UnicodeFromString(string)` could be added
+that would return a `Dafny.ISequence<uint32>` copy instead.
+Migrating an existing codebase should reduce to a simple find-and-replace operation.
+
+Each runtime should also make it possible to wrap an idiomatic representation of a Unicode string
+as the runtime representation of the Dafny `seq<char>` type, without having to make an additional copy. 
+At the time of writing this, only the Java runtime supports this feature,
+via a `StringDafnySequence` class that wraps a `java.lang.String` value and implements the `ISequence<Character>` interface.
+Another similar parallel `UnicodeStringDafnySequence` adaptor can be provided that implements `ISequence<Integer>` instead.
+
+The one wrinkle with this feature is that methods such as `select(i)`, which implement Dafny expressions such as `s[i]`,
+currently take constant time on an `StringDafnySequence`, since a `java.lang.String` also stores its contents as UTF-16
+code units. However, if `UnicodeStringDafnySequence.select(i)` will need to produce the i'th Unicode scalar value instead,
+this will require decoding all characters up to that index, which will take linear time in general instead.
+The solution will be to lazily decode strings into their UTF-32 encoded equivalent (i.e. sequence of scalar values)
```

Comment:
If the point of `StringDafnySequence ` and `UnicodeStringDafnySequence` is to avoid having to make a costly copy of the string, but then accessing individual elements from these sequences is still costly, is there still a point to having them? Would it not be better to force the user to explicitly convert between the native and Dafny string, so they're most aware of the performance cost?

Reply:
I'll add more detail on these wrappers to the actual proposal, but many operations on strings won't need to force decoding because they can operate sequentially on the encoded form, including length, equality, hash code, quantification (i.e. `exists x <- s | 'A' <= x < 'Z'`). Tons of string values will only ever be used as constants and keys in `map`s, and therefore never indexed.

I originally considered caching individual positions, but I assert that most instances of randomly-accessing string positions will hit many such positions in the string: consider parsing workflows, for example. Given calculating the index of any individual character in an encoded string requires scanning from the beginning, and therefore determining the index for all previous characters too, it seems a shame not to hold on to that information. But then in most runtimes indexes into sequences will be 32-bit values anyway, so in some ways it is more memory efficient to just replace the encoded value with the sequence of scalar values as soon as any random access happens. This actually aligns with how Unicode recommends handling encodings anyway: something like UTF-8 is more efficient for storage or APIs, but it's a good idea to decode to UTF-32 if you're doing substantial string processing.

Reply:
Hmm, I'm not sold.  Caching the latest `(index, position)` pair is pretty straightforward and sufficient, no?

Reply:
(And allocation is costly)

Reply:
Yup, if the accesses are in non-decreasing order (and I would expect that even in the parsing use case), that would be sufficient. The downside is having to ensure that mutable state is thread-safe, and that concurrent access will completely destroy the benefits of the cache. 

I'm very comfortable with presenting both solutions as viable options in this proposal, though, and benchmarking on real programs to decide on a solution based on data. Once https://github.com/dafny-lang/dafny/issues/2390 is implemented, it will be very easy to implement both options once in Dafny and test it on each target language. :) I'm less worried about the safety of mutable state in the implementation of immutable values now that I've got a language-agnostic solution in the works: https://github.com/robin-aws/dafny/blob/dafny-runtime-in-dafny/Source/DafnyRuntimeDafny/src/dafnyRuntime.dfy#L296